pkg/agent/cniserver/pod_monitor.go

```diff
@@ -0,0 +1,243 @@
+// Copyright 2024 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cniserver
+
+import (
+	"bytes"
+	"fmt"
+	"sync"
+	"time"
+
+	"antrea.io/libOpenflow/openflow15"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	clientset "k8s.io/client-go/kubernetes"
+	v1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	"antrea.io/antrea/pkg/agent/interfacestore"
+	"antrea.io/antrea/pkg/agent/openflow"
+	agenttypes "antrea.io/antrea/pkg/agent/types"
+	"antrea.io/antrea/pkg/util/channel"
+)
+
+const (
+	podNotReadyTimeInSeconds = 30 * time.Second
+)
+
+var (
+	// retryInterval is the interval to re-install Pod OpenFlow entries if any error happened.
+	// Note, using a variable rather than constant for retryInterval because we may use a shorter time in the
+	// test code.
+	retryInterval = 5 * time.Second
+)
+
+type unreadyPodInfo struct {
+	podName        string
+	podNamespace   string
+	flowsInstalled bool
+	createTime     time.Time
+	ofPort         int32
+}
+
+type podIfaceMonitor struct {
```

Comment:
I may be misunderstanding @tnqn's comment, but the new implementation looks strange to me @wenyingd 

- instead of the `subscribe` / `unsubscribe` method, you are using a workqueue as the interface between the `podConfigurator` and the `podInterfaceMonitor`. I would say this is not a proper interface, as correct workqueue usage really implies that a single component owns the workqueue. So I don't think this is what Quan was suggesting.
- The way I interpret the proposed changes, we would only need `pod_monitor.go` (no need for platform-specific code). The difference would be that either the Linux `podConfigurator` does not subscribe to events, or it ignores them (no-op) - while the Windows `podConfigurator` takes the necessary actions.
- The `podMonitor` still has access to the `ifaceStore`.

It's probably worth it for you to check with @tnqn to get some more guidance

Reply:
> The podMonitor still has access to the ifaceStore. 
This is because I want to use the InterfaceConfig as the single source of the truth of the Pod interface configurations. The `podMonitor` is to waiting for the OpenFlow port of an unready Pod's interface. Once it receives the message, it can update the interface store with the valid of port. In this way,  the podConfigurator can uses the InterfaceConfig to complete the flow installation. 

If I don't use InterfaceStore as the data source, it means I need to introduce another struct to save the OpenFlow port values. Then the podConfigurator (worker) needs to check two sources: 1) interfaceStore to know if the Pod's local configuration exists or not, 2) the new collection for the unready Pod's of port status. This is similar to the previous version that @tnqn suggests to avoid.

> instead of the subscribe / unsubscribe method, you are using a workqueue as the interface between the podConfigurator and the podInterfaceMonitor.
Quan suggests podConfigurator to perform all the operations related with Pod configurations, including flow installation, pod event update, and podMonitor handles the unready interface cache. In this way, except for subscribe/unsubscribe the unready interface, we still need a notification from podMonitor to podConfigurator when the Pod's ofport is ready. otherwise, podConfigurator would not configure the flow in time. I guess you are more prefer to let the podMonitor own the workqueue, and podConfigurator provides a function to be called in the monitor's worker. But even for this, it is actually let the podMonitor perform configuring Pod action, although the action func is provided by struct of podConfigurator. @tnqn may comment it.

>The difference would be that either the Linux podConfigurator does not subscribe to events, or it ignores them (no-op) - while the Windows podConfigurator takes the necessary actions.
In my previous version, I actually make differences in the constructor of podMonitor as platform-specific, it returns nil on Linux and returns a valid podMonitor object on Windows. Then I use "if podMonitor !=nil" in the main workflow (none platform-specific), but Quan sugguests not introduce such if-condition check in the main workflow, but leave the differences in platform-specific implementations. Although it may have a workerqueue run on Linux, but no message sent into it, so it doesn't really impact on the pod configurator.

Reply:
@antoninbas understands the proposal precisely.

`PodConfigurator` could just provide a callback to `PortMonitor` to notify the ready ovsPort and ofPort. I think code like this could separate their responsibilities clearly and avoid unnecessary dependency injection:
```
func newPodConfigurator(
	...
) (*podConfigurator, error) {
	pc := &podConfigurator{
                ...
	}
	pc.ifMonitor = newPodInterfaceMonitor(ofClient, pc.onPortReady)
	return pc, nil
}

func (pc *podConfigurator) onPortReady(ovsPort string, ofPort int) {
	// Update interfacestore if it exists.
	
	// Trigger resync of the target Pod if it exists.
}
```

Ideally, the only job of `PortMonitor` is to waiting for the messages and calling the callback if the port has been subscribed.

Reply:
Another way that maybe even simpler is to just start the goroutine that monitors the channel in configurator itself, which may require less code to fix the problem.
```
func (pc *podConfigurator) Run(stopCh <-chan struct{}) {
	defer pc.queue.ShutDown()
	pc.startEventBroadcaster(stopCh)
	go k8swait.Until(pc.worker, time.Second, stopCh)

	go pc.startMonitoringOFPortStatus(stopCh)

	<-stopCh
}

func (m *podConfigurator) startMonitoringOFPortStatus(stopCh <-chan struct{}) {
	klog.Info("Starting the monitor to wait for new OpenFlow ports")
	statusCh := make(chan *openflow15.PortStatus)
	m.ofClient.SubscribeOFPortStatusMessage(statusCh)
	go wait.Until(func() {
		for {
			select {
			case status := <-m.statusCh:
				klog.V(2).InfoS("Received PortStatus message", "message", status)
				// Update Pod OpenFlow entries only after the OpenFlow port state is live.
				m.processPortStatusMessage(status)
			}
		}
	}, time.Second, stopCh)
}

func (m *podConfigurator) processPortStatusMessage(status *openflow15.PortStatus) {
	// Update Pod OpenFlow entries only after the OpenFlow port state is live.
	if status.Desc.State != openflow15.PS_LIVE {
		return
	}
	ovsPort := string(bytes.Trim(status.Desc.Name, "\x00"))
	ofPort := status.Desc.PortNo

	ifConfig, found := m.ifaceStore.GetInterfaceByName(ovsPort)
	if found {
		// Update interface config with the ofPort.
		ifConfig.OVSPortConfig.OFPort = int32(ofPort)
		m.ifaceStore.UpdateInterface(ifConfig)
	}
	// Add the ovsPort into queue even if there is no related InterfaceConfig in the store, then
	// the worker can delete the unreadyPod if exists.
	m.queue.Add(ovsPort)
}
```

Reply:
got it, updated accordingly.